examples/ch5_example.c

```diff
@@ -0,0 +1,169 @@
+            } else {
+                printf("Hello from job %d\n", *(int *)job->args);
+                // could cause dangling pointer in other threads
+                free(job->args);
+                free(job);
+            }
+        } else {
+            /* To auto run when jobs added, set status to running if job queue is not empty.
+             * As long as the producer is protected */
+            thrd_yield();
+            continue;
+        }
+    };
+}
+
+bool thread_pool_init(thread_pool_t *thrd_pool, size_t size)
+{
+    if (atomic_flag_test_and_set(&thrd_pool->initialezed)) {
+        printf("This thread pool has already been initialized.\n");
+        return false;
+    }
+
+    assert(size > 0);
+    thrd_pool->pool = malloc(sizeof(thrd_t) * size);
+    if (!thrd_pool->pool) {
+        printf("Failed to allocate thread identifiers.\n");
+        return false;
+    }
+
+    // May use memory pool for jobs
+    idle_job_t *idle_job = malloc(sizeof(idle_job_t));
+    if (!idle_job) {
+        printf("Failed to allocate idle job.\n");
+        return false;
+    }
+    // idle_job will always be the first job
+    idle_job->job.args = NULL;
+    idle_job->job.next = &idle_job->job;
+    idle_job->job.prev = &idle_job->job;
+    idle_job->prev = &idle_job->job;
+    thrd_pool->func = worker;
+    thrd_pool->head = idle_job;
+    thrd_pool->state = idle;
+    thrd_pool->size = size;
+
+    for (int i = 0; i < size; i++) {
+        thrd_create(thrd_pool->pool + i, worker, thrd_pool);
+        //TODO: error handling
+    }
+
+    return true;
+}
+
+void thread_pool_destroy(thread_pool_t *thrd_pool)
+{
+    if(atomic_exchange(&thrd_pool->state, cancelled))
+        printf("Thread pool cancelled with jobs still running.\n");
+    for (int i = 0; i < thrd_pool->size; i++) {
+        thrd_join(thrd_pool->pool[i], NULL);
+    }
+    while (thrd_pool->head->prev != &thrd_pool->head->job) {
+        job_t *job = thrd_pool->head->prev->prev;
+        free(thrd_pool->head->prev);
+        thrd_pool->head->prev = job;
+    }
+    free(thrd_pool->head);
+    free(thrd_pool->pool);
+    atomic_fetch_and(&thrd_pool->state, 0);
+    atomic_flag_clear(&thrd_pool->initialezed);
+}
+
+__attribute__((nonnull(2))) bool add_job(thread_pool_t *thrd_pool, void *args)
+{
+    // May use memory pool for jobs
+    job_t *job = malloc(sizeof(job_t));
+    if (!job)
+        return false;
+
+    // unprotected producer
+    job->args = args;
+    job->next = thrd_pool->head->job.next;
+    job->prev = &thrd_pool->head->job;
+    thrd_pool->head->job.next->prev = job;
+    thrd_pool->head->job.next = job;
+    if (thrd_pool->head->prev == &thrd_pool->head->job) {
+        thrd_pool->head->prev = job;
+        // trap worker at idle job
+        thrd_pool->head->job.prev = &thrd_pool->head->job;
+    }
+
+    return true;
+}
+
+int main()
+{
+    thread_pool_t thrd_pool = { .initialezed = ATOMIC_FLAG_INIT };
+    int thread_count = 8;
```

Comment:
Both `job_count` and `thread_count` were meant to be constant, and you can simply specify them as macros. That is, `N_JOBS` and `N_THREADS`.